### PR TITLE
NMS refactor

### PR DESCRIPTION
**Pull request by hgaiser**
This PR refactors NMS, which has a couple of effects:

- The number of outputs for RetinaNet and their meaning changes. Before there was one additional output, `nms_classification`, this changes to `nms_boxes`, `nms_scores`, `nms_labels`. `nms_boxes` is shaped `(300, 4)`, `nms_scores` is shaped `(300,)` and `nms_labels` is shaped `(300,)`. The boxes define the `(x1, y1, x2, y2)` of the detected objects, the scores the classification scores and the labels the predicted labels.
- NMS now supports batch_size > 1.
- The evaluation and visualization code changes accordingly.
- Patch update to the pretrained COCO weights to support the new NMS parameters.

### CODE REVIEW

**Comment by hgaiser** [comment]
Oh also, it speeds up drawing of detections significantly. After https://github.com/fizyr/keras-retinanet/pull/368 the drawing takes ~20msec. With this PR it is ~1msec. It also fixes https://github.com/fizyr/keras-retinanet/issues/334 .

**Comment by Ori226** [comment]
Does this PR also solve #383 ?

**Comment by hgaiser** [comment]
No but it does open the possibility to fix that (I'm working on it at the moment).

File: keras_retinanet/utils/visualization.py [hunk 1]
```diff
@@ -59,25 +59,25 @@ def draw_boxes(image, boxes, color, thickness=2):
...
-        boxes          : A [N, 4] matrix (x1, y1, x2, y2).
-        classification : A [N, num_classes] matrix (cls_1, cls_2, ...).
-        color          : The color of the boxes. By default the color from keras_retinanet.utils.colors.label_color will be used.
-        generator      : (optional) Generator which can map label to class name.
+        image     : The image to draw on.
+        boxes     : A [N, 4] matrix (x1, y1, x2, y2).
+        scores    : A list of N classification scores.
+        labels    : A list of N labels.
+        color     : The color of the boxes. By default the color from keras_retinanet.utils.colors.label_color will be used.
+        generator : (optional) Generator which can map label to class name.
```
Comment (de-vri-es) [inline keras_retinanet/utils/visualization.py]: `score_threshold` isn't documented

File: keras_retinanet/layers/nms.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        other[i] is shaped (max_detections, ...) and contains the filtered other[i] data.
+        In case there are less than max_detections detections, the tensors are padded with -1's.
+    """
+    all_boxes  = []
+    all_scores = []
+    all_labels = []
+    all_other  = []
+
+    # perform per class NMS
+    for c in range(int(classification.shape[1])):
```
Comment (de-vri-es) [inline keras_retinanet/layers/nms.py]: Aren't shapes always ints anyway?

File: keras_retinanet/layers/nms.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    # zero pad the outputs
+    pad_size = keras.backend.maximum(0, max_detections - keras.backend.shape(scores)[0])
+    boxes    = backend.pad(boxes, [[0, pad_size], [0, 0]], constant_values=-1)
+    scores   = backend.pad(scores, [[0, pad_size]], constant_values=-1)
+    labels   = backend.pad(labels, [[0, pad_size]], constant_values=-1)
+    labels   = keras.backend.cast(labels, 'int32')
+    other    = [backend.pad(o, [[0, pad_size]] + [[0, 0] for _ in range(1, len(o.shape))], constant_values=-1) for o in other]
+
+    return [boxes, scores, labels] + other
```
Comment (de-vri-es) [inline keras_retinanet/layers/nms.py]: Looks clean, right amount of comments too :+1:

File: keras_retinanet/layers/nms.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+def nms(boxes, classification, other=[], score_threshold=0.05, max_detections=300, nms_threshold=0.5):
+    """ Run non maximum suppression using the boxes and classification values.
+
+    Args
+        boxes           : Tensor of shape (num_boxes, 4) containing the boxes in (x1, y1, x2, y2) format.
+        classification  : Tensor of shape (num_boxes, num_classes) containing the classification scores.
+        other           : List of tensors of shape (num_boxes, ...) to filter along with the boxes and classification scores.
+        score_threshold : Threshold used to prefilter the boxes with.
+        max_detections  : Maximum number of detections to filter.
```
[the remark this answers is not in the document]
Comment (hgaiser) [reply keras_retinanet/layers/nms.py]: Yep.

File: keras_retinanet/layers/nms.py (hunk 2 above)
Comment (hgaiser) [reply keras_retinanet/layers/nms.py]: They're actually tensors, but in this case they are always representing an int.

Comment (de-vri-es) [reply keras_retinanet/layers/nms.py]: I did say int**s** (plural) ^^ But I'm wondering if the cast to `int` is really needed.

File: keras_retinanet/utils/visualization.py (hunk 1 above)
Comment (hgaiser) [reply keras_retinanet/utils/visualization.py]: Nice catch.

File: keras_retinanet/bin/convert_model.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# Change these to absolute imports if you copy this script outside the keras_retinanet package.
+from .. import models
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser(description='Script for converting a training model to an inference model.')
+
+    parser.add_argument('model', help='The model to convert.')
+    parser.add_argument('converted_model_path', help='Path to save the converted model to.')
```
Comment (de-vri-es) [inline keras_retinanet/bin/convert_model.py]: maybe just name the parameters `model_in` and `model_out` or something? Probably looks better in the help text.

**Comment by hgaiser** [comment]
Updated the PR. The following changes have been made:

- Training works with training models. In other words, the layers required for inference are not added during training (only for evaluation purposes).
- A script is added to convert from a training model to an inference model (which always has the outputs `[boxes, scores, labels, other[0], other[1], ...]`.
- Renamed NonMaximumSuppression layer to FilterDetections layer (as it can be configured to skip NMS now).
- Moved some backbone specific stuff to `keras_retinanet/models/__init__.py`.

File: keras_retinanet/bin/train.py [hunk 6]
```diff
@@ -78,19 +80,14 @@ def create_models(backbone_retinanet, backbone, num_classes, weights, multi_gpu=
...
-        prediction_model = keras.models.Model(inputs=model.inputs, outputs=model.outputs[:2] + [detections])
     else:
-        model            = model_with_weights(backbone_retinanet(num_classes, backbone=backbone, nms=True, modifier=modifier), weights=weights, skip_mismatch=True)
-        training_model   = model
-        prediction_model = model
+        model          = model_with_weights(backbone_retinanet(num_classes, backbone=backbone, modifier=modifier), weights=weights, skip_mismatch=True)
+        training_model = model
+
+    # make prediction model
+    prediction_model = retinanet_bbox(model=model)
```
Comment (de-vri-es) [inline keras_retinanet/bin/train.py]: New model initialization looks good. A lot cleaner :)

File: keras_retinanet/bin/train.py [hunk 7]
```diff
@@ -291,18 +288,7 @@ def check_args(parsed_args):
...
-        from ..models.mobilenet import validate_backbone
-    elif 'vgg' in parsed_args.backbone:
-        from ..models.vgg import validate_backbone
-    elif 'densenet' in parsed_args.backbone:
-        from ..models.densenet import validate_backbone
-    else:
-        raise NotImplementedError('Backbone \'{}\' not implemented.'.format(parsed_args.backbone))
-
-    validate_backbone(parsed_args.backbone)
+    models.validate_backbone(parsed_args.backbone)
```
Comment (de-vri-es) [inline keras_retinanet/bin/train.py]: :+1:

File: keras_retinanet/layers/nms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (de-vri-es) [reply keras_retinanet/layers/nms.py]: It still says to filter.

File: keras_retinanet/layers/filter_detections.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+def filter_detections(boxes, classification, other=[], nms=True, score_threshold=0.05, max_detections=300, nms_threshold=0.5):
+    """ Filter detections using the boxes and classification values.
+
+    Args
+        boxes           : Tensor of shape (num_boxes, 4) containing the boxes in (x1, y1, x2, y2) format.
+        classification  : Tensor of shape (num_boxes, num_classes) containing the classification scores.
+        other           : List of tensors of shape (num_boxes, ...) to filter along with the boxes and classification scores.
+        nms             : Flag to enable/disable non maximum suppression.
+        score_threshold : Threshold used to prefilter the boxes with.
+        max_detections  : Maximum number of detections to filter.
```
Comment (de-vri-es) [inline keras_retinanet/layers/filter_detections.py]: It still says "filter" here.

File: keras_retinanet/layers/filter_detections.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        parallel_iterations = 32,
+        **kwargs
+    ):
+        """ Filters detections using score threshold, NMS and selecting the top-k detections.
+
+        Args
+            nms                 : Flag to enable/disable NMS.
+            nms_threshold       : Threshold for the IoU value to determine when a box should be suppressed.
+            score_threshold     : Threshold used to prefilter the boxes with.
+            max_detections      : Maximum number of detections to keep.
```
Comment (de-vri-es) [inline keras_retinanet/layers/filter_detections.py]: I see you did change it to "keep" here. :+1:

File: keras_retinanet/models/__init__.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+def custom_objects(backbone):
```
Comment (de-vri-es) [inline keras_retinanet/models/__init__.py]: It's good to factor out these operations from the scripts.

If we change this to objects per backbone with these operations it becomes a lot easier to add new backbones though. It would also be a good first step towards backbone plugins. Can be done in a future PR.

File: keras_retinanet/models/densenet.py [hunk 11]
```diff
@@ -17,7 +17,7 @@
 import keras
 from keras.applications.densenet import DenseNet, get_file
 
-from ..models import retinanet
+from . import retinanet
```
Comment (de-vri-es) [inline keras_retinanet/models/densenet.py]: Nice catch.

File: keras_retinanet/models/retinanet.py [hunk 12]
```diff
@@ -25,14 +25,14 @@
...
-    'RegressBoxes'          : layers.RegressBoxes,
-    'NonMaximumSuppression' : layers.NonMaximumSuppression,
-    'Anchors'               : layers.Anchors,
-    'ClipBoxes'             : layers.ClipBoxes,
-    '_smooth_l1'            : losses.smooth_l1(),
-    '_focal'                : losses.focal(),
+    'UpsampleLike'     : layers.UpsampleLike,
+    'PriorProbability' : initializers.PriorProbability,
+    'RegressBoxes'     : layers.RegressBoxes,
+    'FilterDetections' : layers.FilterDetections,
```
Comment (de-vri-es) [inline keras_retinanet/models/retinanet.py]: Also for a future PR, but I think we should prefix out custom objects with `keras_retinanet.`, or maybe even use the full module path and class name. Otherwise it becomes hard (if not impossible) to mix retinanet with other models that use the same names in the same graph.

File: keras_retinanet/models/retinanet.py [hunk 13]
```diff
@@ -330,42 +326,43 @@ def retinanet_bbox(
...
         ```
         [
-            regression, classification, boxes
+            boxes, scores, labels, other[0], other[1], ...
         ]
         ```
     """
-    model = retinanet(inputs=inputs, num_classes=num_classes, **kwargs)
+    if model is None:
+        model = retinanet(anchor_parameters=anchor_parameters, **kwargs)
```
Comment (de-vri-es) [inline keras_retinanet/models/retinanet.py]: How about just passing only `anchor_parameters.num_anchors()` to the `retinanet` function? The less coupling we have the better.

File: keras_retinanet/utils/visualization.py [hunk 14]
```diff
@@ -59,41 +59,42 @@ def draw_boxes(image, boxes, color, thickness=2):
...
 
         # draw labels
-        score   = classification[i, label]
-        caption = (generator.label_to_name(label) if generator else label) + ': {0:.2f}'.format(score)
+        caption = (label_to_name(labels[i]) if label_to_name else labels[i]) + ': {0:.2f}'.format(scores[i])
         draw_caption(image, boxes[i, :], caption)
 
 
-def draw_annotations(image, annotations, color=(0, 255, 0), generator=None):
+def draw_annotations(image, annotations, color=(0, 255, 0), label_to_name=None):
```
Comment (de-vri-es) [inline keras_retinanet/utils/visualization.py]: :tada:

File: keras_retinanet/bin/convert_model.py (hunk 5 above)
Comment (hgaiser) [reply keras_retinanet/bin/convert_model.py]: :+1:

File: keras_retinanet/models/__init__.py (hunk 10 above)
Comment (hgaiser) [reply keras_retinanet/models/__init__.py]: I agree, let's do this in a different PR.

File: keras_retinanet/models/retinanet.py (hunk 13 above)
Comment (hgaiser) [reply keras_retinanet/models/retinanet.py]: :+1:

**Comment by hgaiser** [comment]
Oh yeah, multi-gpu also seems to be working better after this PR. Whether it is working as intended is still to be determined (at the very least the outputs match with what happens without multi-gpu).
